### ISSUE TITLE
Crash receiving a phone call while using AKMicrophone

### ISSUE THREAD

**Issue by stonehouse** [AudioKit/AudioKit#969]
I currently use AudioKit for listening to the microphone and applying FFT analysis. I've discovered that the AVAudioEngine causes a crash if it is started while a phone call is active. If you receive a phone call and then go to an app that is using AudioKit/microphone then the app will crash in the AudioKit.restartEngineAfterRouteChange(_) method when it tries to call self.engine.start().

The only solution I'm aware of for this problem is to check whether a phone call is active before attempting to prepare or start the engine.

I've submitted a radar for this, but I think AudioKit should check if the microphone is in use, and if so, check for active phone calls.

**Comment by aure** [comment]
There have been updates that @warpling implemented for Blackbox, but perhaps those are still in the develop branch.

**Comment by warpling** [comment]
I'm pretty sure this had to do with the phone call triggering the app resigning and an engine interruption which AK hears and then attempts to start the `AVAudioSession` again; it fails when the microphone is active because it's trying to start `AVAudioSession` in a category that it's not allowed to be in (`PlayAndRecord` or `Record`) since the app is not in the foreground and likely doesn't have background audio entitlements. 

This PR stops AK from attempting to restart if it doesn't have the correct entitlement to restart in the category it wants: https://github.com/AudioKit/AudioKit/pull/956

Still in `develop` branch.

**Comment by stonehouse** [comment]
I was able to reproduce the crash with AVFoundation in a separate sample app (which I submitted to Apple in the radar).

```
let engine = AVAudioEngine()
guard let inputNode = engine.inputNode else {
    return
}
			
let bus = 0
inputNode.installTap(onBus: bus, bufferSize: 2048, format: inputNode.inputFormat(forBus: bus), block: { (_,_) in		
})
engine.prepare()
```

If you start a phone call, and run an app with this code, the app will crash immediately.

**Comment by warpling** [comment]
Thank you, personally, for filling that!
Let me know if you get a chance to test the develop branch to see if it fixes the issue too.

**Comment by aure** [comment]
I don't see any actions I can really take on this issue, so I will close it, but it would be great if you still posted updates here about your feedback from Apple on the radar you filed.

**Comment by stonehouse** [comment]
So, I've tried out the develop branch and I saw the same crash again. I've made a PR #1005 demonstrating a fix. I've done something similar in my app to check before we start the AudioKit engine, but something like what I've done in the PR is necessary (IMHO) to prevent a crash on resuming the audio session if a route change notification is received.

I understand if you consider this a low priority problem, but it is currently a major cause of crashes for my app.

**Comment by warpling** [comment]
I was talking with @dave234 and thought this may occur because the Phone has it's own AudioEngine that is holding resources AK's AudioEngine wants when it starts up? This may inform how to check for this before start-up more agnostically and could be tested by creating an AudioEngine that holds audio resources AudioKit needs before attempting to start AudioKit?

**Comment by Rnorback** [comment]
So I'm seeing this crash a lot in my current app. I think people are trying to Transcribe Phone calls.

I've found using the following code:
```
 private func setupMicrophone() {
        let inputFormat = AudioKit.engine.inputNode.inputFormat(forBus: 0)
        AKLog("\nWRONG RATE\nsession rate: \(AVAudioSession.sharedInstance().sampleRate)\ninput rate: \(inputFormat.sampleRate)")
        AKSettings.sampleRate = inputFormat.sampleRate
        AKSettings.channelCount = inputFormat.channelCount
        AKSettings.audioInputEnabled = true
        AKSettings.useBluetooth = true
        AKLog("\nCORRECT RATE\nsession rate: \(AVAudioSession.sharedInstance().sampleRate)\ninput rate: \(inputFormat.sampleRate)\naksettings rate:\(AKSettings.sampleRate)")
        mic = AKMicrophone()
    }
```
Shows the following output during a phone call:
```
WRONG RATE
session rate: 16000.0
input rate: 0.0 (TranscriptionService.swift:setupMicrophone():71)
CORRECT RATE
session rate: 16000.0
input rate: 0.0
aksettings rate:0.0 (TranscriptionService.swift:setupMicrophone():76)

```
So essentially I think what @warpling is saying is correct. I can't seem to manipulate the sample rate of the audio engine during a phone call. I'm worried that this is happening in other ways in my app so I'm going to put a more generalized check for sampleRates than actually looking for a phone call.
